# Improving runtimes of avi-scraper functions using multithreading

I got the low down on possibly improving my runtimes for the HTML parser built to extract avi data. Gonna try some of that out using multithreading 

In [1]:
import re
from datetime import datetime
import requests
import lxml.html as lh
from datetime import date
import time
import pandas as pd
import _thread
from helpers import getAviData
from helpers import getAviData_Optimized
from helpers import get_report_ID_and_process
import threading

I was able to imrove the run time of "getAviData" by creating a multithread of the function responsible for requesting archive pages. Check out the time comparison between getAviData and getAviData_Optimized 

In [2]:
test = ["test1","test2","test3","test4","test5"]
for x in test:
    start = int(time.time())

    # This is the work-horse that gets the data
    getAviData_Optimized(0,x)
    data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
    df = pd.DataFrame(data = data)
    while len(df) != 1043:
        data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
        df = pd.DataFrame(data = data)
        
    finish = int(time.time())
    elapsed_time = (finish - start)/60
    print(x + " took {1:.2f} minutes".format(x, elapsed_time))

test1 took 1.82 minutes
test2 took 1.12 minutes
test3 took 1.38 minutes
test4 took 1.20 minutes
test5 took 1.33 minutes


test1 took 1.82 minutes 

test2 took 1.12 minutes 

test3 took 1.38 minutes 

test4 took 1.20 minutes 

test5 took 1.33 minutes

In [3]:
test = ["test1B","test2B","test3B","test4B","test5B"]
for x in test:
    start = int(time.time())

    # This is the work-horse that gets the data
    getAviData_Optimized(0,x)
    data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
    df = pd.DataFrame(data = data)
    while len(df) != 1043:
        data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
        df = pd.DataFrame(data = data)
        
    finish = int(time.time())
    elapsed_time = (finish - start)/60
    print(x + " took {1:.2f} minutes".format(x, elapsed_time))

test1B took 1.12 minutes
test2B took 1.12 minutes
test3B took 0.95 minutes
test4B took 2.43 minutes
test5B took 1.02 minutes


test1B took 1.12 minutes

test2B took 1.12 minutes

test3B took 0.95 minutes

test4B took 2.43 minutes

test5B took 1.02 minutes

In [2]:
test = ["test1C","test2C","test3C","test4C","test5C"]
for x in test:
    start = int(time.time())

    # This is the work-horse that gets the data
    getAviData(0,x)
    data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
    df = pd.DataFrame(data = data)
    while len(df) != 1043:
        data = pd.read_csv("./"+x+"_zone_0_aviDanger.csv")
        df = pd.DataFrame(data = data)
        
    finish = int(time.time())
    elapsed_time = (finish - start)/60
    print(x + " took {1:.2f} minutes".format(x, elapsed_time))

test1C took 21.95 minutes
test2C took 25.20 minutes
test3C took 25.52 minutes
test4C took 25.70 minutes
test5C took 25.80 minutes


test1C took 21.95 minutes

test2C took 25.20 minutes

test3C took 25.52 minutes

test4C took 25.70 minutes

test5C took 25.80 minutes